src/MySqlConnector/Protocol/Serialization/AuthGSSAPI.cs

```diff
@@ -0,0 +1,237 @@
+				bytesRead += m_readBuffer.Read(buffer, offset, count);
+			}
+			return bytesRead;
+		}
+
+		public override int Read(byte[] buffer, int offset, int count) => ReadAsync(buffer, offset, count, m_cancellationToken).Result;
+
+		public override async Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+		{
+			if (m_writePayloadLength == 0)
+			{
+				// The message header was not read yet.
+				if (count < NegotiateStreamConstants.HeaderLength)
+					// For simplicity, we expect header to be written in one go
+					throw new InvalidDataException("Cannot parse NegotiateStream handshake message header");
+
+				// Parse NegotiateStream handshake header
+				var messageId = buffer[offset+0];
+				var majorProtocolVersion = buffer[offset+1];
+				var minorProtocolVersion = buffer[offset+2];
+				var payloadSizeLow = buffer[offset+4];
+				var payloadSizeHigh = buffer[offset+3];
+
+
+				if (majorProtocolVersion != NegotiateStreamConstants.MajorVersion ||
+					minorProtocolVersion !=  NegotiateStreamConstants.MinorVersion)
+				{
+					throw new FormatException(
+						String.Format("Unknown version of NegotiateStream protocol {0}.{1}, expected {2}.{3}",
+						majorProtocolVersion, minorProtocolVersion,
+						NegotiateStreamConstants.MajorVersion, NegotiateStreamConstants.MinorVersion));
+				}
+				if (messageId != NegotiateStreamConstants.HandshakeDone &&
+					messageId != NegotiateStreamConstants.HandshakeError &&
+					messageId != NegotiateStreamConstants.HandshakeInProgress)
+				{
+					throw new FormatException(
+						String.Format("Invalid NegotiateStream MessageId 0x{0:X2}", messageId));
+				}
+
+				m_writePayloadLength = (int) payloadSizeLow + ((int) payloadSizeHigh << 8);
+				if (messageId == NegotiateStreamConstants.HandshakeDone)
+					m_clientHandshakeDone = true;
+
+				count -= NegotiateStreamConstants.HeaderLength;
+			}
+
+			if (count == 0)
+				return;
+
+			if (count + m_writeBuffer.Length > m_writePayloadLength)
+				throw new InvalidDataException("Attempt to write more than a single message");
+
+			PayloadData payload;
+			if (count <  m_writePayloadLength)
+			{
+				m_writeBuffer.Write(buffer, offset, count);
+				if (m_writeBuffer.Length < m_writePayloadLength)
+					// The message is only partially written
+					return;
+
+				var payloadBytes = m_writeBuffer.ToArray();
+				payload = new PayloadData(new ArraySegment<byte>(payloadBytes, 0, (int) m_writeBuffer.Length));
+				m_writeBuffer.SetLength(0);
+			}
+			else
+			{
+				// full payload provided
+				payload = new PayloadData(new ArraySegment<byte>(buffer, offset, m_writePayloadLength));
+			}
+			await m_serverSession.SendReplyAsync(payload, m_ioBehavior, cancellationToken);
+			// Need to parse NegotiateStream header next time
+			m_writePayloadLength = 0;
+		}
+
+		public override void Write(byte[] buffer, int offset, int count) => WriteAsync(buffer, offset, count, m_cancellationToken).Wait();
```

Comment:
I think this can be `.GetAwaiter().GetResult()` too.

Reply:
Fixed that too.